File: native/tools/dump_parse.cpp
// Stage 2 verification, parsing half.
//
// Prints one deterministic line per field per room. The Python build prints the
// same lines from tools/dump_parse.py, so a plain diff decides whether the port
// is faithful. Nothing here is a test framework; it is the reference numbers.
#include <cstdio>
#include <iostream>
#include <string>

#include "snapir/geometry.hpp"
#include "snapir/parser.hpp"

using namespace snapir;

namespace {

void emit(const std::string& room, const std::string& key, const std::string& value) {
  std::cout << room << '|' << key << '|' << value << '\n';
}

std::string num(double v) {
  char buf[64];
  std::snprintf(buf, sizeof(buf), "%.6f", v);
  return buf;
}

std::string opt(const std::optional<double>& v) { return v ? num(*v) : "None"; }

}  // namespace

int main(int argc, char** argv) {
  if (argc < 2) {
    std::cerr << "usage: dump_parse <folder>\n";
    return 2;
  }

  const Project proj = read_project(argv[1]);
  for (const auto& r : proj.rooms) {
    const std::string& n = r.name;
    emit(n, "outline_source", r.outline_source);
    emit(n, "floor_z", opt(r.floor_z));
    emit(n, "ceiling_z", opt(r.ceiling_z));
    emit(n, "ceiling_height", opt(r.ceiling_height()));
    emit(n, "n_points", std::to_string(r.points.size()));
    emit(n, "n_outline", std::to_string(r.outline.size()));
    emit(n, "n_ceiling", std::to_string(r.ceiling.size()));
    emit(n, "n_openings", std::to_string(r.openings.size()));
    emit(n, "n_controls", std::to_string(r.controls.size()));
    emit(n, "n_segments", std::to_string(r.segments.size()));
    emit(n, "n_links", std::to_string(r.links.size()));
    emit(n, "n_stations", std::to_string(r.stations.size()));
    for (size_t i = 0; i < r.stations.size(); ++i) {
      const auto& s = r.stations[i];
      emit(n, "station[" + std::to_string(i) + "]",
           s.name + " " + num(s.x) + " " + num(s.y) + " " + num(s.z));
    }

    std::vector<Pt> ring;
    for (const auto& p : r.outline) ring.push_back({p.x, p.y});
    emit(n, "area", num(polygon_area(ring)));
    emit(n, "signed_area", num(signed_area(ring)));
    emit(n, "perimeter", num(perimeter(ring)));

    for (size_t i = 0; i < r.outline.size(); ++i) {
      const auto& p = r.outline[i];
      emit(n, "outline[" + std::to_string(i) + "]",
           p.name + " " + num(p.x) + " " + num(p.y) + " " + num(p.z) + " " +
               std::to_string(p.index));
    }
    for (size_t i = 0; i < r.ceiling.size(); ++i) {
      const auto& p = r.ceiling[i];
      emit(n, "ceiling[" + std::to_string(i) + "]", p.name + " " + num(p.z));
    }
    for (size_t i = 0; i < r.openings.size(); ++i) {
      const auto& o = r.openings[i];
      emit(n, "opening[" + std::to_string(i) + "]",
           o.kind + " w=" + num(o.width()) + " sill=" + num(o.sill()) +
               " head=" + num(o.head()) + " L=" + num(o.left.x) + "," + num(o.left.y) +
               " R=" + num(o.right.x) + "," + num(o.right.y));
    }
    // Flights and skirtings, spelled out rather than counted. Two cores can tag
    // the same points and still split them into different flights, and a bare
    // count would compare equal while the geometry differed.
    emit(n, "n_stairs", std::to_string(r.stairs.size()));
    for (size_t i = 0; i < r.stairs.size(); ++i) {
      const auto& s = r.stairs[i];
      std::string names;
      for (size_t k = 0; k < s.points.size(); ++k)
        names += (k ? "," : "") + s.points[k].name;
      emit(n, "stair[" + std::to_string(i) + "]",
           s.kind + " steps=" + std::to_string(s.steps) + " rise=" + num(s.rise()) +
               " going=" + num(s.going()) + " pts=" + names);
    }
    emit(n, "n_pervaz", std::to_string(r.pervaz.size()));
    for (size_t i = 0; i < r.pervaz.size(); ++i) {
      const auto& v = r.pervaz[i];
      emit(n, "pervaz[" + std::to_string(i) + "]",
           v.corner.name + "+" + v.wall.name + " h=" + num(v.height) +
               " d=" + num(v.depth));
    }
    for (size_t i = 0; i < r.issues.size(); ++i)
      emit(n, "issue[" + std::to_string(i) + "]",
           r.issues[i].severity + " " + r.issues[i].code);
    for (const auto& p : r.points)
      emit(n, "role:" + p.name, to_string(p.role));
  }
  return 0;
}

File: tools/dump_parse.py
"""Reference numbers for the C++ port, parsing half.

Prints one line per field per room. native/tools/dump_parse.cpp prints the same
lines, so a plain diff decides whether the port is faithful.
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

# The C++ side writes UTF-8. Windows hands Python a cp1252 stdout by default,
# which cannot encode the dotted capital in the instrument's own name and kills
# the dump before it prints anything - so the two halves could not be compared
# at all on the machine the port is developed on.
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

from snapir.geometry import perimeter, polygon_area, signed_area
from snapir.parser import read_project


def num(v: float) -> str:
    return f"{v:.6f}"


def opt(v) -> str:
    return "None" if v is None else num(v)


def main(folder: str) -> None:
    out = []
    for r in read_project(folder).rooms:
        n = r.name
        def emit(key, value):
            out.append(f"{n}|{key}|{value}")

        emit("outline_source", r.outline_source)
        emit("floor_z", opt(r.floor_z))
        emit("ceiling_z", opt(r.ceiling_z))
        emit("ceiling_height", opt(r.ceiling_height()))
        emit("n_points", len(r.points))
        emit("n_outline", len(r.outline))
        emit("n_ceiling", len(r.ceiling))
        emit("n_openings", len(r.openings))
        emit("n_controls", len(r.controls))
        emit("n_segments", len(r.segments))
        emit("n_links", len(r.links))
        emit("n_stations", len(r.stations))
        for i, s in enumerate(r.stations):
            emit(f"station[{i}]", f"{s.name} {num(s.x)} {num(s.y)} {num(s.z)}")

        ring = [p.xy for p in r.outline]
        emit("area", num(polygon_area(ring)))
        emit("signed_area", num(signed_area(ring)))
        emit("perimeter", num(perimeter(ring)))

        for i, p in enumerate(r.outline):
            emit(f"outline[{i}]",
                 f"{p.name} {num(p.x)} {num(p.y)} {num(p.z)} {p.index}")
        for i, p in enumerate(r.ceiling):
            emit(f"ceiling[{i}]", f"{p.name} {num(p.z)}")
        for i, o in enumerate(r.openings):
            emit(f"opening[{i}]",
                 f"{o.kind} w={num(o.width)} sill={num(o.sill)} head={num(o.head)} "
                 f"L={num(o.left.x)},{num(o.left.y)} R={num(o.right.x)},{num(o.right.y)}")
        # Flights and skirtings, spelled out rather than counted. Two cores can
        # tag the same points and still split them into different flights, and
        # a bare count would compare equal while the geometry differed.
        emit("n_stairs", len(r.stairs))
        for i, s in enumerate(r.stairs):
            emit(f"stair[{i}]",
                 f"{s.kind} steps={s.steps} rise={num(s.rise)} going={num(s.going)} "
                 f"pts={','.join(p.name for p in s.points)}")
        emit("n_pervaz", len(r.pervaz))
        for i, v in enumerate(r.pervaz):
            emit(f"pervaz[{i}]",
                 f"{v.corner.name}+{v.wall.name} h={num(v.height)} d={num(v.depth)}")
        for i, s in enumerate(r.issues):
            emit(f"issue[{i}]", f"{s.severity} {s.code}")
        for p in r.points:
            emit(f"role:{p.name}", p.role.value)

    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main(sys.argv[1])
